analytics-datastore-clickhouse/swarm.sh

```diff
@@ -1,107 +1,113 @@
-  cd "$(dirname "${BASH_SOURCE[0]}")" || exit
-  pwd -P
-)
-readonly COMPOSE_FILE_PATH
-
-ROOT_PATH="${COMPOSE_FILE_PATH}/.."
-readonly ROOT_PATH
-
-. "${ROOT_PATH}/utils/config-utils.sh"
-. "${ROOT_PATH}/utils/docker-utils.sh"
-. "${ROOT_PATH}/utils/log.sh"
-
-main() {
-  if [[ "${ACTION}" == "init" ]] || [[ "${ACTION}" == "up" ]]; then
-    local clickhouse_dev_compose_param=""
-
-    if [[ "${STATEFUL_NODES}" == "cluster" ]]; then
-      log info "Running Analytics Datastore Clickhouse package in Cluster node mode"
-
-      log info "Setting config digests"
-      config::set_config_digests "${COMPOSE_FILE_PATH}/docker-compose.cluster.yml"
-
-      if [[ "${MODE}" == "dev" ]]; then
-        log info "Running Analytics Datastore Clickhouse package in DEV mode"
-        clickhouse_dev_compose_param="-c ${COMPOSE_FILE_PATH}/docker-compose.cluster.dev.yml"
-      fi
+declare ACTION=""
+declare MODE=""
+declare COMPOSE_FILE_PATH=""
+declare ROOT_PATH=""
+declare nodes_mode=""
+declare service_names=()
+
+function init_vars() {
+  ACTION=$1
+  MODE=$2
+
+  COMPOSE_FILE_PATH=$(
+    cd "$(dirname "${BASH_SOURCE[0]}")" || exit
+    pwd -P
+  )
+
+  ROOT_PATH="${COMPOSE_FILE_PATH}/.."
+
+  if [[ "${NODE_MODE}" == "cluster" ]]; then
+    nodes_mode=".${NODE_MODE}"
+    service_names=(
+      "analytics-datastore-clickhouse-01"
+      "analytics-datastore-clickhouse-02"
+      "analytics-datastore-clickhouse-03"
+      "analytics-datastore-clickhouse-04"
+    )
+  else
+    service_names=(
+      "analytics-datastore-clickhouse"
+    )
+  fi
 
-      try "docker stack deploy -c ${COMPOSE_FILE_PATH}/docker-compose.cluster.yml $clickhouse_dev_compose_param instant" "Failed to deploy Analytics Datastore Clickhouse"
+  readonly ACTION
+  readonly MODE
+  readonly COMPOSE_FILE_PATH
+  readonly ROOT_PATH
+  readonly nodes_mode
+  readonly service_names
+}
 
-      docker::await_container_startup analytics-datastore-clickhouse-01
-      docker::await_container_status analytics-datastore-clickhouse-01 Running
+# shellcheck disable=SC1091
+function import_sources() {
+  source "${ROOT_PATH}/utils/docker-utils.sh"
+  source "${ROOT_PATH}/utils/log.sh"
+}
 
-      docker::await_container_startup analytics-datastore-clickhouse-02
-      docker::await_container_status analytics-datastore-clickhouse-02 Running
+function initialize_package() {
+  local clickhouse_dev_compose_param=""
+  if [[ "${MODE}" == "dev" ]]; then
+    log info "Running Analytics Datastore Clickhouse package in DEV mode"
+    clickhouse_dev_compose_param="docker-compose$nodes_mode.dev.yml"
+  else
+    log info "Running Analytics Datastore Clickhouse package in PROD mode"
+  fi
 
-      docker::await_container_startup analytics-datastore-clickhouse-03
-      docker::await_container_status analytics-datastore-clickhouse-03 Running
+  (
+    docker::deploy_service "${COMPOSE_FILE_PATH}" "docker-compose$nodes_mode.yml" "$clickhouse_dev_compose_param"
+    docker::deploy_sanity "${service_names[@]}"
```

Comment:
This deploy sanity function is nice, but I think it should exit on error, which it currently doesn't do. I tested this by adding `- /tmp:/` after line 9 of analytics-datastore-elastic-search/docker-compose.yml. I can't think of errors that we would want to wait on, like, if there is an error that might go away in a few seconds, can you?

Reply:
Actually yes, I faced that one time that's why I fixed it that way, I m sorry I can't remember the exact error, 
but I can like add a condition: if the service faced the same error for 5 times than exit, what do you think ?

Reply:
Hmmm, it seems like the only one to be waiting is deploy_service - that will account for images still being downloaded and that kind of thing. But, deploy_sanity doesn't seem like the kind of thing that should wait. I think we need a second opinion here.